File: Day10-19/10.go
package main

import (
	"encoding/hex"
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

var magic_suffix = []byte{17, 31, 73, 47, 23}

type KnotHash struct {
	hash []byte
	curr int
	skip int
}

func NewKnotHash(size int) KnotHash {
	hash := KnotHash{
		hash: make([]byte, size),
		curr: 0,
		skip: 0,
	}

	for i := 0; i < size; i++ {
		hash.hash[i] = byte(i)
	}
	return hash
}

func knotHash(lengths []byte, state KnotHash) KnotHash {
	size := len(state.hash)
	for _, currLen := range lengths {
		chunkLen := int(currLen)

		// Reverse chunk
		for i := 0; i < chunkLen/2; i++ {
			s := (state.curr + i) % size
			e := (state.curr + chunkLen - i - 1) % size
			tmp := state.hash[s]
			state.hash[s] = state.hash[e]
			state.hash[e] = tmp
		}

		state.curr = (state.curr + (chunkLen + state.skip)) % len(state.hash)
		state.skip++
	}
	return state
}

func knotHashAscii(bytes []byte, rounds int) []byte {
	preparedBytes := append(bytes, magic_suffix...)
	hash := NewKnotHash(256)
	for i := 0; i < 64; i++ {
		hash = knotHash(preparedBytes, hash)
	}

	denseHash := make([]byte, 16)
	for i := 0; i < len(denseHash); i++ {
		xor := byte(0)
		for j := 16 * i; j < 16*i+16; j++ {
			xor ^= hash.hash[j]
		}
		denseHash[i] = xor
	}
	return denseHash
}

func main() {
	dat, _ := ioutil.ReadFile("./10.txt")
	digits := strings.Split(string(dat), ",")
	part1Input := make([]byte, len(digits))
	for idx, digit := range digits {
		intVal, _ := strconv.Atoi(digit)
		part1Input[idx] = byte(intVal)
	}

	part1Hash := NewKnotHash(256)
	knotHash(part1Input, part1Hash)
	part1Result := int(part1Hash.hash[0]) * int(part1Hash.hash[1])
	fmt.Printf("Part 1: %d\n", part1Result)

	part2DenseHash := knotHashAscii(dat, 64)
	fmt.Printf("Part 2: %s\n", hex.EncodeToString(part2DenseHash))
}

File: Day10-19/10.py
import numpy as np

magic_suffix = [17, 31, 73, 47, 23]


def knot_hash(inputs, size=256, curr=0, skip=0, knot=None):
    if knot is None:
        knot = np.arange(0, size)
    for i in inputs:
        if curr + i < size:
            chunk = knot[curr:curr+i]
        else:
            chunk = np.append(knot[curr:], knot[:i-(size-curr)])
        chunk = chunk[::-1]
        if curr + i < size:
            knot[curr:curr+i] = chunk
        else:
            knot[curr:] = chunk[:size-curr]
            knot[:i-(size-curr)] = chunk[size-curr:]
        curr = (curr + (i + skip)) % size
        skip += 1
    return knot, curr, skip


def part1(inputs, size=256):
    lengths, _, _ = knot_hash(inputs, size=size)
    return lengths[0] * lengths[1]


def part2(input_str):
    lengths = [ord(i) for i in input_str] + magic_suffix
    sparse_hash = None
    curr, skip = 0, 0
    for _ in range(64):
        sparse_hash, curr, skip = knot_hash(
            lengths, curr=curr, skip=skip, knot=sparse_hash
        )
    dense_hash = [0 for _ in range(16)]
    for i in range(16):
        dense_hash[i] = np.bitwise_xor.reduce(sparse_hash[16*i:16*i+16])
    return(bytes(dense_hash).hex())

if __name__ == '__main__':
    # assert part1([3, 4, 1, 5], size=5) == 12
    # assert part2('') == 'a2582a3a0e66e6e86e3812dcb672a272'
    # assert part2('AoC 2017') == '33efeb34ea91902bb2f59c9920caa6cd'

    with open("10.txt") as f:
        raw = f.read().strip()
    lengths = [int(i) for i in raw.split(',')]
    product = part1(lengths)
    print("Part 1: {}".format(product))

    dense_hash = part2(raw)
    print("Part 2: {}".format(dense_hash))
